feat!: fx and interscheme implementation

kleyow opened the pull request:
**Changes:**
- Currency Conversion (FX)
  - Added fxTransfer API implementation
  - Updated notification handler to deal with FX transfers
- Performance optimisations
  - Option to send audit and trace events directly to Kafka instead of using sidecars
- Interscheme
  - Parameterised switch id
  - Proxy cache implementation using redis
  - FORWARDED message production in notification handler

elnyry-sam-k commented on .circleci/config.yml:
```diff
@@ -1,976 +1,11 @@
-            - test-integration
-            - license-scan
-            - image-scan
+      - build/workflow:
```
Does this cover all the existing steps? example license scans, image scans? It needs to at least cover every existing CI step..

Please confirm

kalinkrustev replied:
Yes, but you can also check in CircleCI to confirm.
For example https://app.circleci.com/pipelines/github/mojaloop/ml-api-adapter/1158/workflows/96e01bfd-61c8-4fff-b360-c329cb5bbfee

elnyry-sam-k replied:
Thanks @kalinkrustev

elnyry-sam-k commented:
Thanks Kevin for the PR and Shashi for the review.

Have some general comments, please review along with the CI config comment above:

1. Please add header files to the newly added source files
2. Need to address - TODOs; If they're not required for this release, please add a follow-up issue on Mojaloop GitHub with the details but if they are required for this release, they need to be addressed
3. Need to standardize the ulid regex validation (from line: transferId: Joi.string().pattern(/^[0-9a-f]{8}-[0-9a-f]{4}-[1-7][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$|^[0-9A-HJKMNP-TV-Z]{26}$/).required().description('Id of transfer').label('Transfer Id must be in a valid GUID/ULID format.'), in file src/api/transfers/routes.js . Either a function from a standard library or if unavailable, something in sdk-standard-components, because it'll apply for all "CorrelationId" types used


Sample header:

```
/*****
 License
 --------------
 Copyright © 2020-2024 Mojaloop Foundation
 The Mojaloop files are made available by the Mojaloop Foundation under the Apache License, Version 2.0 (the "License") and you may not use these files except in compliance with the License. You may obtain a copy of the License at

 http://www.apache.org/licenses/LICENSE-2.0

 Unless required by applicable law or agreed to in writing, the Mojaloop files are distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied. See the License for the specific language governing permissions and limitations under the License.

 Contributors
 --------------
 This is the official list of the Mojaloop project contributors for this file.
 Names of the original copyright holders (individuals or organizations)
 should be listed with a '*' in the first column. People who have
 contributed from an organization can be listed under the organization
 that actually holds the copyright for their contributions (see the
 Mojaloop Foundation for an example). Those individuals should have
 their names indented and be marked with a '-'. Email address can be added
 optionally within square brackets <email>.

 * Mojaloop Foundation
 - Name Surname <name.surname@mojaloop.io>


 --------------

 ******/
```

shashi165 commented:
The integration tests are failing locally for this branch

kalinkrustev commented:
> The integration tests are failing locally for this branch

@shashi165 any errors that you can share?

kleyow commented:
https://github.com/mojaloop/ml-api-adapter/pull/562

@elnyry-sam-k 
>Please add header files to the newly added source files

Added source header files to newly added files in PR.

> Need to address - TODOs; If they're not required for this release, please add a follow-up issue on Mojaloop GitHub with the details but if they are required for this release, they need to be addressed

I've removed the added todo's in the PR. They seemed minor.

>Need to standardize the ulid regex validation (from line: transferId: Joi.string().pattern(/^[0-9a-f]{8}-[0-9a-f]{4}-[1-7][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$|^[0-9A-HJKMNP-TV-Z]{26}$/).required().description('Id of transfer').label('Transfer Id must be in a valid GUID/ULID format.'), in file src/api/transfers/routes.js . Either a function from a standard library or if unavailable, something in sdk-standard-components, because it'll apply for all "CorrelationId" types used.

Is the regex not adequate for now?
`^[0-9a-f]{8}-[0-9a-f]{4}-[1-7][0-9a-f]{3}-[89ab][0-9a-f]{3}-[0-9a-f]{12}$` represents uuids and `^[0-9A-HJKMNP-TV-Z]{26}$` represents ulids.

Also in the `minor/iso` branch that we're hoping will shortly follow we have ripped out `joi` in favor of using the standard central shared openapi validator with the standardized openapi specs.

https://github.com/mojaloop/ml-api-adapter/commit/1385e1e4b4d99313b438480f98c05b698894ab96
It's not a small piece of work to port to this PR.

Let me know if we can ignore this in favor of not duplicating work. 

@shashi165 any update on the errors?
Did you make sure to run `export ENDPOINT_URL=http://localhost:4545/notification` before running the harness and int tests?

elnyry-sam-k commented:
Ok, all good with me here.. happy to resolve the comments based on follow-up actions. thanks @kleyow

elnyry-sam-k reviewed (approved):
Looks good, thank you!
